File: 133-clone-graph/clone-graph.java
class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;

        // Map to store the original node to its clone
        HashMap<Node, Node> map = new HashMap<>();
        Queue<Node> queue = new LinkedList<>();

        // Initialize BFS
        queue.add(node);
        map.put(node, new Node(node.val));

        while (!queue.isEmpty()) {
            Node currNode = queue.poll();

            // Iterate through neighbors
            for (Node neighbor : currNode.neighbors) {
                if (!map.containsKey(neighbor)) {
                    // Clone the neighbor if not already cloned
                    map.put(neighbor, new Node(neighbor.val));
                    queue.add(neighbor);
                }
                // Add the cloned neighbor to the current node's neighbors
                map.get(currNode).neighbors.add(map.get(neighbor));
            }
        }

        return map.get(node);
    }
}

File: 133-clone-graph/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) 
    {
        if(node == NULL)
        return NULL;

        unordered_map<Node*, Node*> map;
        map[node] = new Node(node->val);

        queue<Node*> q;
        q.push(node);
        while(!q.empty())
        {
            Node* currNode = q.front();
            q.pop();

            for(Node* nbr : currNode->neighbors)
            {
                // clone the neighbor only if its not cloned
                if(!map.count(nbr))
                {
                    map[nbr] = new Node(nbr->val);
                    q.push(nbr);
                }

                // Add the cloned neighbor to the current node's neighbors
                map[currNode]->neighbors.push_back(map[nbr]);
            }
        }
        return map[node];
    }
};